### PR TITLE
Updated error returns on rmw_take

### PR DESCRIPTION
**Pull request by Lobotuerk**
Signed-off-by: lobotuerk <jtlorente@ekumenlabs.com>

### CODE REVIEW

**Comment by Lobotuerk** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12301/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7273/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10017/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12199/)

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 1]
```diff
@@ -2515,10 +2515,19 @@ static rmw_ret_t rmw_take_int(
...
-  RET_NULL(subscription);
-  RET_WRONG_IMPLID(subscription);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    taken, "taken handle is null",
+    return RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    ros_message, "ros message handle is null",
+    return RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    subscription, "subscription info is null",
```
Comment (Blast545) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:2525]:
```suggestion
    subscription, "subscription handle is null",
```

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 2]
```diff
@@ -2515,10 +2515,19 @@ static rmw_ret_t rmw_take_int(
...
+    return RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    subscription handle,
```
Comment (Blast545) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:2528]: This looks odd to me
```suggestion
    subscription handle,
```

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 1 above)
Comment (Lobotuerk) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:2525]: Updated

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 2 above)
Comment (Lobotuerk) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:2528]: Could you explain this one?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 3]
```diff
@@ -2515,10 +2515,22 @@ static rmw_ret_t rmw_take_int(
   const rmw_subscription_t * subscription, void * ros_message,
   bool * taken, rmw_message_info_t * message_info)
 {
-  RET_NULL(taken);
-  RET_NULL(ros_message);
-  RET_NULL(subscription);
-  RET_WRONG_IMPLID(subscription);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    taken, "taken handle is null",
+    return RMW_RET_INVALID_ARGUMENT);
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:2520]: @Lobotuerk hmm, `taken` is not a handle. I believe you want to use [`RMW_CHECK_ARGUMENT_FOR_NULL`](https://github.com/ros2/rmw/blob/efe3f2bf932aabd70e875228e39b843103f2c4b3/rmw/include/rmw/error_handling.h#L107). Same elsewhere.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 4]
```diff
@@ -2515,10 +2515,22 @@ static rmw_ret_t rmw_take_int(
...
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    subscription, "subscription handle is null",
+    return RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    message_info, "message info is null",
+    return RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    subscription handle,
+    subscription->implementation_identifier, eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION)
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:2533]: @Lobotuerk nit: trailing semicolon?

Comment (Lobotuerk) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Changed

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 3 above)
Comment (Lobotuerk) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:2520]: One of the implementations used that message with the error, I'm trying to be consistent between implementations.

Comment (Lobotuerk) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:2520]: Updated for RMW_CHECK_ARGUMENT_FOR_NULL

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 5]
```diff
@@ -2557,20 +2569,37 @@ static rmw_ret_t rmw_take_seq(
...
+    subscription, RMW_RET_INVALID_ARGUMENT);
   RET_WRONG_IMPLID(subscription);
 
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    subscription handle,
+    subscription->implementation_identifier, eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  if (count == 0u) {
+    RMW_SET_ERROR_MSG("count cant be 0");
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:2591]: @Lobotuerk nit: 
```suggestion
    RMW_SET_ERROR_MSG("count cannot be 0");
```

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 6]
```diff
@@ -2557,20 +2569,37 @@ static rmw_ret_t rmw_take_seq(
...
+  RMW_CHECK_ARGUMENT_FOR_NULL(
+    subscription, RMW_RET_INVALID_ARGUMENT);
   RET_WRONG_IMPLID(subscription);
 
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    subscription handle,
+    subscription->implementation_identifier, eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  if (count == 0u) {
```
Comment (hidmic) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp:2590]: @Lobotuerk nit: 
```suggestion
  if (0u == count) {
```
per style guide.

**Review by hidmic — APPROVED** [review]
LGTM but for a few minor nits

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 5 above)
Comment (Lobotuerk) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:2591]: Updated

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 6 above)
Comment (Lobotuerk) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp:2590]: Updated

**Comment by Lobotuerk** [comment]
Updated jobs:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12363/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7332/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10075/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12275/)

**Review by hidmic — APPROVED** [review]

**Review by Blast545 — APPROVED** [review]
LGTM

**Comment by jacobperron** [comment]
This change introduced several test failures downstream ([for example](https://ci.ros2.org/job/ci_linux/12374/testReport/)).

Locally, if I revert this PR then tests start passing again.

@Lobotuerk For future, we should run CI for all packages when making changes to an RMW implementation.

**Comment by hidmic** [comment]
> For future, we should run CI for all packages when making changes to an RMW implementation.

This also means that the corresponding test in `test_rmw_implementation` is not sufficient.

**Pull request merged**